### PR TITLE
HCM: protect against removal of critical response headers by a filter chain.

### LINKED ISSUE #13756: Gracefully handle removal of critical headers by filters
It happened on several occasions that buggy filters modify or remove headers critical to Envoy's internals (or other filters in the chain). Envoy should check if required request or response headers are still present and valid after invoking filter's `decodeHeaders()` or `encodeHeaders()` method. If a filter has modified or deleted critical header the request should be aborted with 500 response.

Critical request header are:

- :host

Critical response headers are:

- :status

### PR DESCRIPTION
**Pull request by mathetake**
This is the rework of the previous PR (https://github.com/envoyproxy/envoy/pull/15658).

Commit Message: HCM: protect against removal of critical response headers by a filter chain.
Additional Description: Previously, Envoy hasn't been guarded from the removal of critical response headers like :status. This happens when there's misbehaving filter chan, and especially this protection is important when users run their own extensions through Wasm. Resolves #13756 and ref: #15487.

*Notes*

- For the request path, Envoy is already guarded by HeaderUtility::checkRequiredHeaders and its error is gracefully handled in the onPoolReady.

- The following is the current crash path with uncaught exception caused by removed :status header after filter chains:

https://github.com/envoyproxy/envoy/blob/5c51bf16660db4713f8879faf0d34e2075739273/source/common/http/conn_manager_impl.cc#L1481

https://github.com/envoyproxy/envoy/blob/5c51bf16660db4713f8879faf0d34e2075739273/source/common/http/conn_manager_impl.cc#L777

https://github.com/envoyproxy/envoy/blob/5c51bf16660db4713f8879faf0d34e2075739273/source/common/http/utility.cc#L455

- In the previous PR, I implemented this guard in the codec and make the codec return the status, but it's going to make the implementation much more complex than I expected (You can see in the size of the diff in https://github.com/envoyproxy/envoy/pull/15658), plus introduce a lot of complexity in the code base (there were even unresolved comment about its direction). So I came back to the first solution there as opposed to the comment by @asraa https://github.com/envoyproxy/envoy/pull/15658#discussion_r601131466).

Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

### CODE REVIEW

File: source/common/http/header_utility.cc
Comment (asraa) [inline source/common/http/header_utility.cc:348]: could you share code with `Utility::getResponseStatusNoThrow`?

**Review by asraa — COMMENTED** [review]
i think my vote for making encodeHeaders (although larger change) return a status is that we won't need to check elsewhere in case some other extension point besides filters break status (i think that was the comment on my request side PR). but maybe that should change when that will be a problem

File: docs/root/configuration/http/http_conn_man/response_code_details.rst
Comment (asraa) [inline docs/root/configuration/http/http_conn_man/response_code_details.rst:30]: -> e.g. overflown status?

**Comment by mathetake** [comment]
> we won't need to check elsewhere in case some other extension point besides filters break status

Thanks, that makes sense.. but yeah we could expand this check when actually we see other extension points break the status header.

File: test/common/http/utility_test.cc
Comment (asraa) [inline test/common/http/utility_test.cc:1487]: add response header tests?
(very minor nit, but i'm not sure what NG stands for, can you make this slightly more verbose?)

**Review by asraa — DISMISSED** [review]
The code otherwise looks good to me, deferring to a senior maintainer for any opinions on the approach
@envoyproxy/senior-maintainers

File: test/common/http/utility_test.cc
Comment (mathetake) [reply test/common/http/utility_test.cc]: https://github.com/envoyproxy/envoy/pull/16745/commits/7953e305afe2be0d312223b87ab4025e18927965 done, thanks!

File: docs/root/configuration/http/http_conn_man/response_code_details.rst
Comment (mathetake) [reply docs/root/configuration/http/http_conn_man/response_code_details.rst]: done: https://github.com/envoyproxy/envoy/pull/16745/commits/7953e305afe2be0d312223b87ab4025e18927965

File: source/common/http/header_utility.cc
Comment (alyssawilk) [inline source/common/http/header_utility.cc:347]: can we have a code comment why this is here, rather than returning from the codec encode headers?

File: source/common/http/header_utility.h
Comment (alyssawilk) [inline source/common/http/header_utility.h:207]: I hear your point about calling this from the codec being really messy.
optional!: Could we at least as part of this PR make the codecs not throw/crash if they get bad data a well?  I think we'd prefer to serialize 0 if status code is not present to crashing though Matt may disagree =P

**Review by alyssawilk — COMMENTED** [review]
thanks for fixing this.

LGTM modulo comment request.

Non-blocking comments, fixing this crash is good, but if we want to allow semi-trusted wasm, I think we should do the full suite of header checks that we do in the codec after wasm runs, ensure all headers only have valid characters, we don't allow smuggling by having transfer encoding and content length, relative urls don't start with ../ etc? Is there a plan and/or tracking issue for that?

File: source/common/http/header_utility.h
Comment (mathetake) [reply source/common/http/header_utility.h:207]: In order to do so, we should eliminate all the usage of `getResponseStatus` and use `getResponseStatusNoThrow` instead? e.g.
- https://github.com/envoyproxy/envoy/blob/5c51bf16660db4713f8879faf0d34e2075739273/source/common/http/http2/codec_impl.cc#L336
- https://github.com/envoyproxy/envoy/blob/5c51bf16660db4713f8879faf0d34e2075739273/source/common/http/http1/codec_impl.cc#L373-L375

I think fixing this is worth a separate PR and discussion, so can I open an issue for now and discuss there?

File: source/common/http/header_utility.cc
Comment (mathetake) [reply source/common/http/header_utility.cc:347]: done (https://github.com/envoyproxy/envoy/pull/16745/commits/46d285a2903de395ff768bc0b2dc20686b97720c)

**Comment by mathetake** [comment]
> but if we want to allow semi-trusted wasm, I think we should do the full suite of header checks that we do in the codec after wasm runs, ensure all headers only have valid characters, we don't allow smuggling by having transfer encoding and content length, relative urls don't start with ../ etc? Is there a plan and/or tracking issue for that?

yeah that's true and this is somewhat related to the comment by @PiotrSikora https://github.com/envoyproxy/envoy/issues/15487#issuecomment-802185648, and yes, we should do Wasm specific check in addition to this Envoy-wide protection. I will open an issue for tracking this. Thanks!

**Review by alyssawilk — APPROVED** [review]

File: source/common/http/header_utility.h
Comment (alyssawilk) [reply source/common/http/header_utility.h:207]: SGTM, and as it's optional don't feel obliged to tackle it (though it would be nice!)
I think it's largely a part of removing exceptions from the data plane, which I think @chaoqin-li1123 is looking at

**Pull request merged**
